src/osp/Active/SysNewton.h

```diff
@@ -105,28 +107,55 @@ class SysNewton : public IDynamicSystem
     
     void update_world(ActiveScene& rScene);
 
-    constexpr ActiveScene& get_scene() noexcept { return m_scene; }
-
     static ACompNwtWorld* try_get_physics_world(ActiveScene &rScene);
 
     /**
      * Used to find which rigid body an entity belongs to. This will keep
      * looking up the tree of parents until it finds a rigid body.
      *
-     * @param rScene [in] ActiveScene containing ent
-     * @param ent    [in] ActiveEnt with ACompHierarchy and rigid body ancestor
+     * @param rScene     [in] ActiveScene containing ent
+     * @param ent        [in] ActiveEnt with ACompHierarchy and rigidbody ancestor
+     * @param pTransform [out] A matrix to store the relative orienation of ent
+                              with respect to the ancestor
      *
      * @return Pair of {level-1 entity, pointer to ACompNwtBody found}. If
      *         hierarchy error, then {entt:null, nullptr}. If no ACompNwtBody
      *         component is found, then {level-1 entity, nullptr}
      */
     static std::pair<ActiveEnt, ACompNwtBody*> find_rigidbody_ancestor(
-            ActiveScene& rScene, ActiveEnt ent);
+            ActiveScene& rScene, ActiveEnt ent, Matrix4* pTransform = nullptr);
+
+    /**
+     * Helper function for a SysMachine to fill a rigid body member
+     *
+     * Used by systems that need to fetch a rigidbody. Consider a childEntity
+     * which possesses a machine that interacts with the rigidbody (e.g. a
+     * thruster). Such a machine will store an ActiveEnt& member that refers to
+     * the rigidbody, which will be uninitialized at first. This member will be
+     * passed to parentRigidBody. If the reference points to a valid entity, and
+     * that entity has a rigidbody component, the rigidbody and its
+     * transformation are simply returned. If not, SysNewton attempts to find
+     * the rigidbody ancestor of childEntity. If both of these cases fail, the
+     * function returns nullptrs.
+     * 
+     * @param rScene          [in] The scene to search
+     * @param childEntity     [in] An entity whose rigidbody ancestor is sought
+     * @param rParentRigidBody [in/out] A reference to a machine member which
+     *                          either contains the entity owning the rigidbody
+     *                          ancestor component, or if not, said entity will
+     *                          be found and assigned as the value of the member
+     * 
+     * @return Pair of {pointer to found ACompNwtBody, pointer to ACompTransform
+     *         of the ACompNwtBody entity}. If either component can't be found,
+     *         returns {nullptr, nullptr}
+     */
```

Comment:
Yeah I struggled to document this function because its behavior is kind of complicated. Look at its actual definition, you'll recognize it from MachineRocket. It takes an in/out `ActiveEnt&` which receives the machine's handle on the rigid body ancestor. If it's null, the function finds it and fills it in, otherwise it just retrieves the rb component.